### PR TITLE
Add a new blog

### PR DESCRIPTION
**Pull request by husiyu**
@amyangfei PTAL

### CODE REVIEW

File: dm-source-code-reading-10.md [hunk 1]
```diff
@@ -0,0 +1,123 @@
+---
+title: DM 源码阅读系列文章（十）测试框架的实现
+author: ['杨非']
+date: 2019-07-23
+summary: 本篇文章将从质量保证的角度来介绍 DM 测试框架的设计和实现，探讨如何通过多维度的的测试方法保证 DM 的正确性和稳定性。 
+tags: ['DM 源码阅读','社区']
+---
+
+本文为 DM 源码阅读系列文章的第十篇，之前的文章已经详细介绍过 DM 数据同步各组件的实现原理和代码解析，相信大家对 DM 的实现细节已经有了深入的了解。本篇文章将从质量保证的角度来介绍 DM 测试框架的设计和实现，探讨如何通过多维度的的测试方法保证 DM 的正确性和稳定性。
```
Comment (amyangfei) [inline dm-source-code-reading-10.md]:
```suggestion
本文为 DM 源码阅读系列文章的第十篇，之前的文章已经详细介绍过 DM 数据同步各组件的实现原理和代码解析，相信大家对 DM 的实现细节已经有了深入的了解。本篇文章将从质量保证的角度来介绍 DM 测试框架的设计和实现，探讨如何通过多维度的测试方法保证 DM 的正确性和稳定性。
```

File: dm-source-code-reading-10.md [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+## 测试体系
+
+DM 完整的测试体系包括以下四个部分：
+
+### 1. 单元测试
+
+主要用于测试每个 go 模块和具体函数实现的正确性，测试用例的编写依照 go 单元测试的标准，跟随代码一起发布。具体测试用例使用的 [pingcap/check](https://github.com/pingcap/check) 工具包，该工具包是在 go 原生测试工具基础上进行的扩展，[按照 suite 分组进行测试](https://github.com/pingcap/check/blob/67f458068fc864dabf17e38d4d337f28430d13ed/run.go#L98-L131)，提供包括更丰富的检测语法糖，并行测试，序列化测试在内的一些扩展特性。单元测试的设计出发点是白盒测试，测试用例中通过尽可能明确的测试输入得到期望的测试输出。
```
Comment (amyangfei) [inline dm-source-code-reading-10.md]:
```suggestion
主要用于测试每个 go 模块和具体函数实现的正确性，测试用例的测试方法依照 go 单元测试的标准，跟随代码一起发布。具体测试用例编写使用 [pingcap/check](https://github.com/pingcap/check) 工具包，该工具包是在 go 原生测试工具基础上进行的扩展，[按照 suite 分组进行测试](https://github.com/pingcap/check/blob/67f458068fc864dabf17e38d4d337f28430d13ed/run.go#L98-L131)，提供包括更丰富的检测语法糖、并行测试、序列化测试在内的一些扩展特性。单元测试的设计出发点是白盒测试，测试用例中通过尽可能明确的测试输入得到期望的测试输出。
```

File: dm-source-code-reading-10.md [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+### 2. 集成测试
+
+用于测试各个组件之间交互的正确性和完整数据同步流程的正确性，完整的 [测试用例集合和测试工具在项目代码的 tests 目录](https://github.com/pingcap/dm/tree/7cba6d21d78dd16e9ab159e9c0300efcbdeb1e4a/tests) 发布。集成测试首先自定义了一些 [DM 基础测试工具集](https://github.com/pingcap/dm/tree/7cba6d21d78dd16e9ab159e9c0300efcbdeb1e4a/tests/_utils)，包括启动 DM 组件，生成、导入测试数据，检测同步状态、上下游数据一致性等 bash 脚本，每个测试用例是一个完整的数据同步场景，通过脚本实现数据准备、启动 DM 集群、模拟上游数据输入、特定异常和恢复、数据同步校验等测试流程。集成测试的设计出发点是确定性的模拟测试场景，为了能够确定性的模拟一些特定的同步场景，为此我们还引入了 [failpoint](https://github.com/pingcap/failpoint) 来注入测试、控制测试流程， 以及 [trace](https://github.com/pingcap/dm/tree/7cba6d21d78dd16e9ab159e9c0300efcbdeb1e4a/pkg/tracing) 机制来更准确地获取程序内存状态、辅助控制测试流程，具体的实现细节会在后文详细介绍。
+
+### 3. 破坏性测试
+
+真实的软件运行环境中会遇到各种各样的问题，包括各类硬件故障、网络延迟和隔离、资源不足等等。DM 在数据同步过程中也同样会遇到这些问题，借助于 [PingCAP 内部的自动混沌测试平台 schrodinger](https://thenewstack.io/chaos-tools-and-techniques-for-testing-the-tidb-distributed-newsql-database/)，我们设计了多个破坏性测试用例，包括在同步过程中随机 kill DM-worker 节点，同步过程中重启部分 DM-worker 节点，分发不兼容 DDL 语句等测试场景。这一类测试的关注点是在各类破坏性操作之后数据同步能否正常恢复以及验证在这些场景下数据一致性的保证，测试用例通常以黑盒的形式去运行，并且长期、反复地进行测试。
```
Comment (amyangfei) [inline dm-source-code-reading-10.md]:
```suggestion
真实的软件运行环境中会遇到各种各样的问题，包括各类硬件故障、网络延迟和隔离、资源不足等等。DM 在数据同步过程中也同样会遇到这些问题，借助于 [PingCAP 内部的自动化混沌测试平台 schrodinger](https://thenewstack.io/chaos-tools-and-techniques-for-testing-the-tidb-distributed-newsql-database/)，我们设计了多个破坏性测试用例，包括在同步过程中随机 kill DM-worker 节点，同步过程中重启部分 DM-worker 节点，分发不兼容 DDL 语句等测试场景。这一类测试的关注点是在各类破坏性操作之后数据同步能否正常恢复以及验证在这些场景下数据一致性的保证，测试用例通常以黑盒的形式去运行，并且长期、反复地进行测试。
```

File: dm-source-code-reading-10.md [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+| 破坏性测试 | 黑盒测试，随机数据，随机触发的固定类型外部扰动 | 系统在异常场景下的稳定性和正确性 | 在内部测试平台长期、反复运行 | 对已有确定输入测试的补充，增加测试输入的不确定性，通过未知、随机的外部扰动发现系统潜在的问题 |
+| 长期稳定性测试 | 黑盒测试，确定性的同步场景，随机数据负载 | 系统长期运行的稳定性和正确性 | 在内部 K8s 集群长期运行 | 补充集成测试的场景，测试系统在更高负载、更长运行时间内的表现 |
+
+## 测试 case 与测试工具的实现 
+
+### 1. 在单元测试中进行 mock
+
+我们在单元测试运行过程中希望尽量减少外部环境或内部组件的依赖，譬如测试 relay 模块时我们并不希望从上游的 MySQL 拉取 binlog，或者测试到下游的一些数据库读写操作并不希望真正部署一个下游 TiDB，这时候我们就需要对测试 case 进行适当的 mock。在单元测试中针对不同的场景采用了多种 mock 方案。接下来我们选取几种具有代表性的方案进行介绍。
+
+#### mock golang interface
```
Comment (amyangfei) [inline dm-source-code-reading-10.md]:
```suggestion
#### Mock golang interface
```

**Comment by husiyu** [comment]
@YiniXu9506 PTAL

**Pull request merged**